### DISCUSSION TITLE
进度更新 Progress update

### DISCUSSION THREAD

**Discussion by NineDayCC** [NineDayCC/HeadTracker_ESP32#2] [General]
我会不定期在这里更新进度和计划。有什么想法也可以在这里提出来，催更也可以来这里哈哈。
I will update the progress and plans here irregularly. You can also share your thoughts here and feel free to urge me for updates.

**Comment by NineDayCC** [comment]
目前计划先继续使用原来的蓝牙，用esp-idf重新写好先。然后再看一下怎么兼容elrs的无线头追功能，实现直连elrs。
顺便解释一下，不用 zephyrOS 改用 esp idf 是因为 zephyr 目前还有好多功能不支持，会有PPM抖动的问题，而且不支持esp now。

The current plan is to continue using the original Bluetooth function and rewrite it with esp-idf. Then, we'll manage to integrate the wireless head tracking feature of elrs and establish a direct connection with elrs.
By the way, the reason for switching from zephyrOS to esp-idf is that zephyrOS currently lacks support for many functions and may experience PPM jitter issues, and it does not support esp now.

**Comment by NineDayCC** [comment]
第一版pre-release已经准备好了。现在已经能用蓝牙连接来实现无线头追了。

1. 支持PPM输出，且无zephyrOS版本会出现的抖动现象。
2. 暂不支持SBUS，后续版本追加
3. 延迟稍大，由于使用的是蓝牙连接，延迟稍微有点感觉
4. 上电就自动连接，不需要匹配。但是多个此版本头追同时使用时，注意分开上电，否则可能连到别人的头追
5. 短按回中，长按保持当前位置（头追动，输出不动）

预告一下后面的计划：
1. 连接改用esp now协议，降低延迟
2. 直连elrs。兼容elrs的无线头追功能。[https://github.com/ExpressLRS/ExpressLRS/pull/2060](url)
3. 增加上位机调参，校准功能

如果大家有什么想到可以增加的功能，欢迎留言评论！


The first version pre-release is ready. Now it is possible to achieve wireless head tracking through Bluetooth connection.

1. Support for PPM output, without the jitter issue that may occur in ZephyrOS versions.
2. SBUS is not currently supported, but will be added in future versions.
3. Slightly higher latency due to the use of Bluetooth connection.
4. Automatically connects upon power-up, no need for pairing. However, when multiple devices with this version of head tracking are used simultaneously, make sure to power them on separately to avoid connecting to someone else's head tracker.
5. Short press recenters the head, long press maintains the current position (headtreacker moves, output remains static).

Here's a preview of our future plans:
1. Switch to the esp now protocol for lower latency in the connection.
2. Directly connect to elrs. Compatible with elrs wireless head tracking feature. [https://github.com/ExpressLRS/ExpressLRS/pull/2060](url)
3. Add parameter adjustment and calibration functions through a computer interface.

If anyone has any suggestions for additional features, please feel free to leave a comment!

**Comment by NineDayCC** [comment]
很久没有在这里给大家更新进展了，最近多了很多朋友的关注，也完成了一些阶段性的进展，在这里给大家介绍一下项目的进展吧。

### 硬件方面
硬件目前有两版硬件，分别是 **HT_Nano** 和 **HT_Nano_V2** 。接收端目前只有一版，可以通用。

**HT_Nano** 使用 **ESP32-PICO-V3** 作为主控，IMU 使用 **ICM-42688P**。
已知的问题主要：

- 有发热大（采用LDO，电源效率低）
- 制作困难（螺柱焊接困难，容易虚焊）

 **HT_Nano_V2** 使用 **ESP32-C3** 作为主控，IMU 使用 **ICM-42688P**。
V2 版本主要更新了：

- DCDC供电，降低发热
- 优化制作方法，提高制作良率
- 更换主控芯片，降低成本和方便硬件 layout
- 优化了射频布局，理论上通信距离会更远（待测试）
- 增加了USB切换开关，可以使用 usb 直连电脑，进行烧录和测试。
- 使用触摸芯片实现触摸功能（ESP32C3没有触摸IO口，需要独立的芯片实现触摸功能）

V2 版本刚完成，尚未正式开源。后面会先开源原理图，待资料整理好后再全部开源。

### 软件方面
当前最新版本为 HT_V1.0.0  #6
1. 发射端与接收端的通信现在采用 ESP-NOW 。
2. 现在支持使用网页进行 OTA 。
3. 回中按键采用触摸，点按回中，长按固定输出。
4. 蜂鸣器提示音。
5. typec 直连供电，可以插入到如 DJI G2 、DJI N3 等，带有支持 PD 输出的 typec 口的头追上。
6. 双 typec 口，使用时不影响手机连接头追，进行投屏等操作。

如果你有什么想要了解的，欢迎与我交流！

-----

Long time no update! Recently, we’ve gained many new followers and made some milestone progress. Here’s a project status recap:

### Hardware Updates
We currently have two hardware versions: **HT_Nano** and **HT_Nano_V2**. The receiver remains universal for both.

**HT_Nano**
MCU: **ESP32-PICO-V3**
IMU: **ICM-42688P**
Known Issues:

- Overheating (due to LDO power supply inefficiency).
- Manufacturing challenges (soldering standoffs is prone to cold joints).

**HT_Nano_V2**
MCU: **ESP32-C3**
IMU: **ICM-42688P**
Key Improvements:

- Switched to DCDC power supply, reducing heat.
-Optimized manufacturing process for higher yield.
-Changed MCU to lower cost and simplify PCB layout.
-Enhanced RF circuit design for theoretically longer range (pending tests).
-Added a USB toggle switch for direct PC connection (flashing/testing).
-Implemented touch functionality via a dedicated chip (ESP32-C3 lacks touch-capable GPIOs).

V2 is freshly completed but not yet open-sourced. We’ll release the schematics first, followed by full documentation once finalized.

### Software Updates
Latest Version: HT_V1.0.0 #6

1. ESP-NOW for TX/RX communication.
2. Web-based OTA update support.
3. Touch-based center button: Tap: Re-center. Hold: Lock output.
4. Buzzer audio feedback.
5. Type-C PD power delivery (compatible with DJI G2/N3 headsets’ Type-C PD ports).
6. Dual Type-C ports (allows simultaneous headtracking and phone connectivity for screen mirroring).

Feel free to reach out with questions or suggestions!

**Reply by KhangNguyen1307** [reply]
what a good news to hearXD

**Comment by NineDayCC** [comment]
最近我做好了几个 V2 版本的头追，已经寄出给朋友测试了。

制作过程中我发现 V2 版本虽然优化过制作过程，但是需要手工的部分还是比较多，贴片成本较贵。

所以我正在**更新一版 SE 版本的头追**。新版本会去掉大部分的非必要功能，只保留最基础的功能。力求最小化制作成本，方便批量制作。

新版本会去掉如 双 Type-C，蜂鸣器，机械按键等。仅保留一个typec公头，一个触摸回中按键，和两个LED。

软件方面，我已经大致了解了 elrs 无线头追的连接过程，后续软件上会优先增加 elrs 无线头追的支持。

------
Recently, I completed several V2 versions of head trackers and have already sent them to friends for testing.  

During the production process, I noticed that although the V2 version has optimized the manufacturing steps, it still requires a significant amount of manual work, and the cost of the PCB assembly is relatively high.  

Therefore, I am currently working on an **updated SE version of the head tracker**. The new version will remove most non-essential features, retaining only the most basic functionalities. The goal is to minimize production costs and facilitate bulk manufacturing.  

The new version will eliminate features such as dual Type-C ports, a buzzer, and mechanical buttons, keeping only one Type-C male connector, one touch-based centering button, and two LEDs.  

As for software, I have largely figured out the connection process for ELRS wireless head tracking. Moving forward, the software will prioritize adding support for ELRS wireless head tracking.

**Comment by NineDayCC** [comment]
最新 SE 版硬件已经初步测试完成了。

这版硬件更加容易制作，取消了所有机械按键。固件已经适配，后续将会更新文档。

详细的可以查看此合并 #8 
***
The latest SE Edition hardware has passed initial testing.

This version is easier to manufacture, as all physical buttons have been removed. The firmware has been adapted accordingly, and documentation will be updated later.

For details, check out this merge: #8
